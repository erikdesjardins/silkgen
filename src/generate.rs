use crate::analyze::{Extents, PixelKind};
use crate::sizes::{KicadDim, KicadPos, PixelDim, PixelPos};
use image::{DynamicImage, GenericImageView};
use rand::Rng;
use std::io;
use std::io::Write;
use std::ops::Neg;

pub fn output_file(
    name: &str,
    image: DynamicImage,
    pixel_pitch: KicadDim,
    mut r: impl Rng,
    mut w: impl Write,
) -> Result<(), io::Error> {
    let image = image.into_luma_alpha8();

    log::info!(
        "Total image dimensions: {} x {}",
        image.width(),
        image.height()
    );

    let extents = Extents::from_image(&image);

    log::info!("Extent of significant pixels: {:?}", extents);
    log::info!("Center of significant pixels: {:?}", extents.center());

    sexpr(&mut w, "footprint", |w| {
        w.write_all(b"\"")?;
        w.write_all(name.replace('\"', "").as_bytes())?;
        w.write_all(b"\"\n")?;

        // prelude
        sexpr(w, "version", |w| w.write_all(b"20220630"))?;
        sexpr(w, "generator", |w| w.write_all(b"silkgen"))?;
        sexpr(w, "layer", |w| w.write_all(b"F.Silkscreen"))?;
        sexpr(w, "tedit", |w| w.write_all(b"0"))?;
        sexpr(w, "attr", |w| {
            w.write_all(b"board_only exclude_from_pos_files exclude_from_bom")
        })?;

        // text references
        sexpr(w, "fp_text", |w| {
            w.write_all(b"reference \"G***\" (at 0 0) (layer F.Fab)\n")?;
            w.write_all(b"(effects (font (size 1.524 1.524) (thickness 0.3)))\n")?;
            tstamp(w, &mut r)
        })?;
        sexpr(w, "fp_text", |w| {
            w.write_all(b"value \"LOGO\" (at 0.75 0) (layer F.Fab) hide\n")?;
            w.write_all(b"(effects (font (size 1.524 1.524) (thickness 0.3)))\n")?;
            tstamp(w, &mut r)
        })?;

        // pixels
        for (x, y, pixel) in GenericImageView::pixels(&image) {
            let layers: &[_] = match PixelKind::from_pixel(pixel) {
                PixelKind::Transparent => continue,
                PixelKind::Light => &["F.SilkS"],
                PixelKind::Dark => &["F.Cu", "F.Mask"],
            };
            for layer in layers {
                draw_pixel(
                    w,
                    &mut r,
                    PixelPos {
                        x: PixelDim(x),
                        y: PixelDim(y),
                    },
                    &extents,
                    pixel_pitch,
                    layer,
                )?;
            }
        }

        Ok(())
    })
}

fn draw_pixel(
    w: &mut impl Write,
    r: &mut impl Rng,
    top_left: PixelPos,
    extents: &Extents,
    pixel_pitch: KicadDim,
    layer: &str,
) -> Result<(), io::Error> {
    sexpr(w, "fp_poly", |w| {
        sexpr(w, "pts", |w| {
            // shift by 1 since we base positions on the top left of the pixel
            let center = extents.center() + PixelPos::X1 + PixelPos::Y1;
            // place all 4 corners of the polygon
            let kicad_pos = |pos: PixelPos| KicadPos {
                x: neg_if(pos.x < center.x, pixel_pitch * (pos.x.abs_diff(center.x))),
                y: neg_if(pos.y < center.y, pixel_pitch * (pos.y.abs_diff(center.y))),
            };
            xy(w, kicad_pos(top_left))?;
            xy(w, kicad_pos(top_left + PixelPos::X1))?;
            xy(w, kicad_pos(top_left + PixelPos::X1 + PixelPos::Y1))?;
            xy(w, kicad_pos(top_left + PixelPos::Y1))?;
            Ok(())
        })?;
        sexpr(w, "layer", |w| w.write_all(layer.as_bytes()))?;
        sexpr(w, "width", |w| w.write_all(b"0"))?;
        sexpr(w, "fill", |w| w.write_all(b"solid"))?;
        tstamp(w, r)
    })
}

fn xy(w: &mut impl Write, pos: KicadPos) -> Result<(), io::Error> {
    sexpr(w, "xy", |w| write!(w, "{} {}", pos.x, pos.y))
}

fn tstamp(w: &mut impl Write, r: &mut impl Rng) -> Result<(), io::Error> {
    sexpr(w, "tstamp", |w| {
        let uuid = uuid::Builder::from_random_bytes(r.gen()).into_uuid();
        w.write_all(uuid.to_string().as_bytes())
    })
}

fn sexpr<W: Write, R>(
    w: &mut W,
    name: &str,
    f: impl FnOnce(&mut W) -> Result<R, io::Error>,
) -> Result<R, io::Error> {
    w.write_all(b"(")?;
    w.write_all(name.as_bytes())?;
    w.write_all(b" ")?;
    let r = f(w)?;
    w.write_all(b")\n")?;
    Ok(r)
}

fn neg_if<T: Neg<Output = T>>(cond: bool, x: T) -> T {
    if cond {
        -x
    } else {
        x
    }
}
