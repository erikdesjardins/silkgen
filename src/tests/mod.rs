use crate::generate;
use rand::rngs::StdRng;
use rand::SeedableRng;

#[test]
fn basic() {
    insta::assert_snapshot!(run_against(include_bytes!("basic.png")));
}

fn run_against(file: &[u8]) -> String {
    let image = image::load_from_memory(file).unwrap();

    let pixel_pitch = "1mm".parse().unwrap();

    let rng = StdRng::from_seed(*b"1234567890abcdefghijklmnopqrstuv");

    let mut out = Vec::new();

    generate::output_file("testname", image, pixel_pitch, rng, &mut out).unwrap();

    String::from_utf8(out).unwrap()
}
