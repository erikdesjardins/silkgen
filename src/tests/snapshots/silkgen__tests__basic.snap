---
source: src/tests/mod.rs
expression: "run_against(include_bytes!(\"basic.png\"))"
---
(footprint "testname"
(version 20220630)
(generator silkgen)
(layer F.Silkscreen)
(tedit 0)
(attr board_only exclude_from_pos_files exclude_from_bom)
(fp_text reference "G***" (at 0 0) (layer F.Fab)
(effects (font (size 1.524 1.524) (thickness 0.3)))
(tstamp c45669fa-23d3-4b15-a7c8-d69b555815e1)
)
(fp_text value "LOGO" (at 0.75 0) (layer F.Fab) hide
(effects (font (size 1.524 1.524) (thickness 0.3)))
(tstamp e8692f66-44d9-4022-bcff-21cb4616bb87)
)
(fp_poly (pts (xy -1 -1)
(xy 0 -1)
(xy 0 0)
(xy -1 0)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp 082a15f5-203b-440c-9ef1-b46049776530)
)
(fp_poly (pts (xy 0 -1)
(xy 1 -1)
(xy 1 0)
(xy 0 0)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp df5ea2bf-9422-42e0-b8be-9882e9db0a75)
)
(fp_poly (pts (xy 0 -1)
(xy 1 -1)
(xy 1 0)
(xy 0 0)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp 40233e12-fc59-4519-9f01-16e978915da5)
)
(fp_poly (pts (xy -1 0)
(xy 0 0)
(xy 0 1)
(xy -1 1)
)
(layer F.Cu)
(width 0)
(fill solid)
(tstamp eab1ba90-f6ba-4511-8675-00d9416b041a)
)
(fp_poly (pts (xy -1 0)
(xy 0 0)
(xy 0 1)
(xy -1 1)
)
(layer F.Mask)
(width 0)
(fill solid)
(tstamp be8de91f-f686-460f-9f5e-5162fe705b44)
)
(fp_poly (pts (xy 0 0)
(xy 1 0)
(xy 1 1)
(xy 0 1)
)
(layer F.SilkS)
(width 0)
(fill solid)
(tstamp d33b933a-5654-44bf-bc1b-3deeb6a7a900)
)
)

